### ISSUE TITLE
android virtual device fails to identify camera device 

### ISSUE THREAD

**Issue by asluborski98** [google/mediapipe#3165]
Hello I am trying to run hands solution with ADV on pixel XL API 30. I want to get access to camera but when I try to use camera it gives this error. 

```
E/EGL_emulation: eglQueryContext 32c0  EGL_BAD_ATTRIBUTE
E/EGL_emulation: tid 10572: eglQueryContext(1902): error 0x3004 (EGL_BAD_ATTRIBUTE)
I/CameraManagerGlobal: Connecting to camera service
D/CameraRepository: Added camera: 0
I/Camera2CameraInfo: Device Level: INFO_SUPPORTED_HARDWARE_LEVEL_LIMITED
D/UseCaseAttachState: Active and attached use case: [] for camera: 0
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.google.mediapipe.apps.hands, PID: 10349
    java.lang.IllegalArgumentException: No available camera can be found.
        at androidx.camera.core.CameraSelector.filter(CameraSelector.java:100)
        at androidx.camera.lifecycle.ProcessCameraProvider.bindToLifecycle(ProcessCameraProvider.java:389)
        at androidx.camera.lifecycle.ProcessCameraProvider.bindToLifecycle(ProcessCameraProvider.java:275)
        at com.google.mediapipe.components.CameraXPreviewHelper.lambda$startCamera$4$CameraXPreviewHelper(CameraXPreviewHelper.java:317)
        at com.google.mediapipe.components.-$$Lambda$CameraXPreviewHelper$XxlOsze8Y8Q9eLsR5HjR1U3b8zU.run(Unknown Source:16)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
I/Process: Sending signal. PID: 10349 SIG: 9
```
I am running gradle solution and the AVD camera app has no problem opening up my webcam camera, just with the mediapipe app so I assume there is something in the code I have to change or it is not possible to run with AVD?

I am using latest version of Android studio w/ Android 11.0

I have my Emulated Performance graphics set to Software - GLES 2.0

My front and back camera are both set to Webcam0

My SD card is studio-managed and has 512 MB allocated to it. 

I am running Linux OS and can confirm my webcam has necessary drivers installed and works in other applications. 

Any pointers as to what could be causing this issue?

**Comment by asluborski98** [comment]
OK - So I got it kind of working by changing front camera to emulated and back camera to webcam 0.

Mediapipe app now recognizes my camera and for a second I see myself but then it crashes with this error

```
/UseCaseAttachState: All use case: [androidx.camera.core.Preview-f9d36866-9a9d-4db0-867a-773c450aa561170607875] for camera: 1
D/DeferrableSurface: New surface in use[total_surfaces=1, used_surfaces=1](androidx.camera.core.SurfaceRequest$2@c126fe}
D/DeferrableSurface: use count+1, useCount=1 androidx.camera.core.SurfaceRequest$2@c126fe
D/CaptureSession: Opening capture session.
D/CaptureSession: Attempting to send capture request onConfigured
D/CaptureSession: Issuing request for session.
W/Gralloc4: allocator 3.x is not supported
D/CaptureSession: CameraCaptureSession.onConfigured() mState=OPENED
D/CaptureSession: CameraCaptureSession.onReady() OPENED
D/ExternalTextureConv: Created output texture: 2 width: 720 height: 960
D/ExternalTextureConv: Created output texture: 3 width: 720 height: 960
D/ExternalTextureConv: Created output texture: 4 width: 720 height: 960
I/native: I20220314 22:15:08.669728  2617 jni_util.cc:41] GetEnv: not attached
I/native: I20220314 22:15:08.835422  2619 jni_util.cc:41] GetEnv: not attached
I/native: I20220314 22:15:08.841735  2604 jni_util.cc:41] GetEnv: not attached
D/GlThread: Stopping GL thread ExternalTextureConverter
    
    --------- beginning of crash
E/AndroidRuntime: FATAL EXCEPTION: ExternalTextureConverter
    Process: com.google.mediapipe.apps.hands, PID: 1731
    java.lang.RuntimeException: glViewport: GL error: 0x501
        at com.google.mediapipe.glutil.ShaderUtil.checkGlError(ShaderUtil.java:175)
        at com.google.mediapipe.glutil.GlThread.bindFramebuffer(GlThread.java:117)
        at com.google.mediapipe.components.ExternalTextureConverter$RenderThread.updateOutputFrame(ExternalTextureConverter.java:485)
        at com.google.mediapipe.components.ExternalTextureConverter$RenderThread.renderNext(ExternalTextureConverter.java:389)
        at com.google.mediapipe.components.ExternalTextureConverter$RenderThread.lambda$onFrameAvailable$0$ExternalTextureConverter$RenderThread(ExternalTextureConverter.java:338)
        at com.google.mediapipe.components.-$$Lambda$ExternalTextureConverter$RenderThread$IXJjtE68JMV2CMJNQcFYhTUqhj8.run(Unknown Source:4)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at com.google.mediapipe.glutil.GlThread.run(GlThread.java:141)
E/BufferQueueProducer: [SurfaceTexture-1-1731-0](id:6c300000000,api:4,p:407,c:1731) dequeueBuffer: BufferQueue has been abandoned
E/BufferQueueProducer: [SurfaceTexture-1-1731-0](id:6c300000000,api:4,p:407,c:1731) queueBuffer: BufferQueue has been abandoned
E/BufferQueueProducer: [SurfaceTexture-1-1731-0](id:6c300000000,api:4,p:407,c:1731) queueBuffer: BufferQueue has been abandoned
D/CaptureSession: CameraCaptureSession.onReady() OPENED
I/native: I20220314 22:15:09.015316  2616 jni_util.cc:41] GetEnv: not attached
I/Process: Sending signal. PID: 1731 SIG: 9
```

**Comment by sureshdagooglecom** [comment]
Hi @asluborski98 ,could you provide steps to reproduce this issue.

**Comment by asluborski98** [comment]
Yes so I just built the gradle app according to official Mediapipe instructions and run it on the above emulator device settings on Pixel XL. 

I then run the app on the emulator which opens the menu. Picture and video buttons and inference work fine - it is only when I hit camera it freezes and the above error log comes up for real-time. 

I was able to succesfully use it on physical device through adb which is OK for me - but still would be nice to find out what the issue is with emulator
